src/util/system.cpp

```diff
@@ -874,6 +874,39 @@ std::string ArgsManager::GetChainName() const
     return GetArg("-chain", CBaseChainParams::MAIN);
 }
 
+void ArgsManager::logArgsPrefix(
+    const std::string& prefix,
+    const std::string& section,
+    const std::map<std::string, std::vector<util::SettingsValue>>& args) const
+{
+    std::string sectionStr = section.empty() ? "" : "[" + section + "] ";
+    for (const auto& arg : args) {
+        for (const auto& value : arg.second) {
+            if (FlagsOfKnownArg(arg.first) & DONT_LOG) {
+                LogPrintf("%s %s%s=****\n", prefix, sectionStr, arg.first);
+            } else if (value.isBool()) {
+                LogPrintf("%s %s%s=%d\n", prefix, sectionStr, arg.first, value.get_bool());
+            } else if (value.isStr()) {
+                std::string valueStr = value.get_str();
+                if (valueStr.size()) {
+                    LogPrintf("%s %s%s=%s\n", prefix, sectionStr, arg.first, valueStr);
+                } else {
+                    LogPrintf("%s %s%s\n", prefix, sectionStr, arg.first);
+                }
+            }
```

Comment:
In commit "merge with 'Deduplicate settings' commit" (d95fdc78c965708866156898588221a8d983396f)

Would suggest:

```c++
if (FlagsOfKnownArg(arg.first) & DONT_LOG) {
    LogPrintf("%s %s%s=****\n", prefix, sectionStr, arg.first);
} else {
    LogPrintf("%s %s%s=%s\n", prefix, sectionStr, arg.first, value.write());
}
```

This is simpler than formatting code in the PR, more future proof, and also less ambiguous about negated options.

Would also suggest squashing commits.

Reply:
@ryanofsky, thanks, I like that much better. I implemented your suggestion (with a small difference, only a single call to `LogPrintf()`), rebased, squashed, force-pushed (eb973ff916f88ed36b59ae6466ec16dd30811215). The formatting is slightly different now, but I like that the code is much simpler and the output is more consistent (this is from one of the functional tests):
```
2019-11-26T16:09:54.048744Z [init] config-file arg: regtest="1"
2019-11-26T16:09:54.048755Z [init] config-file arg: [regtest] bind="127.0.0.1"
2019-11-26T16:09:54.048765Z [init] config-file arg: [regtest] discover="0"
2019-11-26T16:09:54.048774Z [init] config-file arg: [regtest] dnsseed="0"
2019-11-26T16:09:54.048784Z [init] config-file arg: [regtest] fallbackfee="0.0002"
2019-11-26T16:09:54.048793Z [init] config-file arg: [regtest] keypool="1"
2019-11-26T16:09:54.048802Z [init] config-file arg: [regtest] listenonion="0"
2019-11-26T16:09:54.048812Z [init] config-file arg: [regtest] port="11068"
2019-11-26T16:09:54.048822Z [init] config-file arg: [regtest] printtoconsole="0"
2019-11-26T16:09:54.048833Z [init] config-file arg: [regtest] rpcport="16068"
2019-11-26T16:09:54.048843Z [init] config-file arg: [regtest] server="1"
2019-11-26T16:09:54.048854Z [init] config-file arg: [regtest] shrinkdebugfile="0"
2019-11-26T16:09:54.048865Z [init] config-file arg: [regtest] upnp="0"
2019-11-26T16:09:54.048874Z [init] command-line arg: datadir="/tmp/bitcoin_func_test_wmar_kd1/node0"
2019-11-26T16:09:54.048884Z [init] command-line arg: debug=""
2019-11-26T16:09:54.048895Z [init] command-line arg: debugexclude="libevent"
2019-11-26T16:09:54.048904Z [init] command-line arg: debugexclude="leveldb"
2019-11-26T16:09:54.048914Z [init] command-line arg: logthreadnames=""
2019-11-26T16:09:54.048924Z [init] command-line arg: logtimemicros=""
2019-11-26T16:09:54.048934Z [init] command-line arg: uacomment="testnode0"
```
I like the way negative arguments are displayed; as you say, they are less ambiguous. There are no examples of such from above, but I ran `src/bitcoind -noreindex` and the following appears in `debug.log`:
```
2019-11-26T17:00:24Z command-line arg: reindex=false
```
I added a unit test for this too. Thank you for taking the time to review!